tests/test_api_requests.py

```diff
@@ -322,6 +233,1467 @@ async def parse(self, response):
+    [
+        ("default_params", "zyte_api"),
+        (
+            "automap_params",
+            "zyte_api_automap",
+        ),
+    ],
+)
+def test_default_params_immutability(arg_key, meta_key, setting, meta):
+    """Make sure that the merging of Zyte API parameters from the *arg_key*
+    _get_api_params parameter with those from the *meta_key* request metadata
+    key does not affect the contents of the setting for later requests."""
+    request = Request(url="https://example.com")
+    request.meta[meta_key] = meta
+    default_params = copy(setting)
+    _get_api_params(
+        request,
+        **{
+            **GET_API_PARAMS_KWARGS,
+            arg_key: default_params,
+        },
     )
+    assert default_params == setting
+
+
+def _test_automap(global_kwargs, request_kwargs, meta, expected, warnings, caplog):
+    request = Request(url="https://example.com", **request_kwargs)
+    request.meta["zyte_api_automap"] = meta
+    with caplog.at_level("WARNING"):
+        api_params = _get_api_params(
+            request,
+            **{
+                **GET_API_PARAMS_KWARGS,
+                **global_kwargs,
+                "transparent_mode": True,
+            },
+        )
+    assert api_params == expected
+    if warnings:
+        for warning in warnings:
+            assert warning in caplog.text
+    else:
+        assert not caplog.records
+
+
+@pytest.mark.parametrize(
+    "meta,expected,warnings",
+    [
+        # If no other known main output is specified in meta, httpResponseBody
+        # is requested.
+        ({}, {"httpResponseBody": True, "httpResponseHeaders": True}, []),
+        (
+            {"unknownMainOutput": True},
+            {
+                "httpResponseBody": True,
+                "httpResponseHeaders": True,
+                "unknownMainOutput": True,
+            },
+            [],
+        ),
+        # If httpResponseBody is unnecessarily requested in meta, a warning is
+        # logged.
+        (
+            {"httpResponseBody": True},
+            {"httpResponseBody": True, "httpResponseHeaders": True},
+            ["do not need to set httpResponseBody to True"],
+        ),
+        # If other main outputs are specified in meta, httpRequestBody is not
+        # set.
+        (
+            {"browserHtml": True},
+            {"browserHtml": True, "httpResponseHeaders": True},
+            [],
+        ),
+        (
+            {"screenshot": True},
+            {"screenshot": True},
+            [],
+        ),
+        (
+            {"browserHtml": True, "screenshot": True},
+            {"browserHtml": True, "httpResponseHeaders": True, "screenshot": True},
+            [],
+        ),
+        # If no known main output is specified, and httpResponseBody is
+        # explicitly set to False, httpResponseBody is unset and no main output
+        # is added.
+        (
+            {"httpResponseBody": False},
+            {},
+            [],
```

Comment:
Should we log a warning that the user is not sending any meaningful payload to Zyte API?

Reply:
I don’t think so.

Zyte API will currently fail with a bad request error, which achieves the same goal, I think. And if in the future this becomes a valid request, then the warning would be wrong.

On the other hand, if the way they make this a valid request in the future is to make `"httpResponseBody": True` the default in this case, then the automapped parameters (none) would seemingly contradict user intent (`"httpResponseBody": False`). But I think the only scenario where a user sets `"httpResponseBody": False` is one where the user includes some other (currently unknown) parameter that indicates a different output, and I don’t expect Zyte API to ever make `"httpResponseBody": True` implicit in such cases.